### ISSUE TITLE
Couple Ginkgo tests are failing on NERSC Perlmutter

### ISSUE THREAD

**Issue by pwxy** [ginkgo-project/ginkgo#1292]
I tried running the ginkgo tests on Perlmutter with both gcc and nvidia compiler builds.

For gcc 11.2.0 compiler with cuda 11.7 builds, "make test" gives:

```
99% tests passed, 1 tests failed out of 282

Total Test time (real) = 373.87 sec

The following tests FAILED:
250 - test/matrix/matrix_cuda (Failed)
```

I've attached the "gcc-LastTest.log" file, which showed that while the "make test" output to the screen stated 1 test failed, LastTest.log shows that actually 4 tests failed:

```
[  FAILED  ] 4 tests, listed below:
[  FAILED  ] Matrix/CsrWithSparselibStrategy.SpMVIsEquivalentToRef, where TypeParam = CsrWithSparselibStrategy
[  FAILED  ] Matrix/CsrWithSparselibStrategy.AdvancedSpMVIsEquivalentToRef, where TypeParam = CsrWithSparselibStrategy
[  FAILED  ] Matrix/CsrWithSparselibStrategy.MixedSpMVIsEquivalentToRef, where TypeParam = CsrWithSparselibStrategy
[  FAILED  ] Matrix/CsrWithSparselibStrategy.MixedAdvancedSpMVIsEquivalentToRef, where TypeParam = CsrWithSparselibStrategy
```

For the xgc work, the test/matrix/matrix_cuda failure isn't an issue because it's for CSR format, and we use ELL instead.  But I had seen the errors were pretty large.

My cmake line:

```
cmake \
-DCMAKE_CXX_COMPILER=g++ \
-DGINKGO_COMPILER_FLAGS=-pedantic \
-DCMAKE_INSTALL_PREFIX=${install_dir} \
-DGINKGO_CUDA_ARCHITECTURES=Ampere \
-DGINKGO_BUILD_OMP=OFF \
-DGINKGO_BUILD_MPI=OFF \
-DCMAKE_BUILD_TYPE:STRING=RELEASE \
-DGINKGO_BUILD_TESTS=on \
-DGINKGO_BUILD_BENCHMARKS=off \
-DGINKGO_BUILD_EXAMPLES=off \
-DGINKGO_BUILD_HWLOC=off \
..
```

I then tried ginkgo tests with nvidia 22.9 compiler and cuda 11.7, and quite a few more tests failed (5 additional tests failed):

```
98% tests passed, 6 tests failed out of 282

Total Test time (real) = 447.98 sec

The following tests FAILED:
         39 - reference/test/factorization/lu_kernels (Failed)
         90 - reference/test/solver/cgs_kernels (Failed)
        133 - core/test/base/math (Failed)
        249 - test/matrix/dense_kernels_cuda (Failed)
        250 - test/matrix/matrix_cuda (Failed)
        253 - test/preconditioner/jacobi_kernels_cuda (Failed)
```

I've attached the "nvidia-LastTest.log" file, which showed that while the "make test" output to the screen stated 6 tests failed, LastTest.log shows that actually many more than 6 tests failed.

Looks like those other test failures were barely missing the tolerance (e.g. error was 1e-13 and tolerance was 1e-14), so I had wondered if the tolerance was too tight. 

[gcc-CMakeCache.txt](https://github.com/ginkgo-project/ginkgo/files/10911541/gcc-CMakeCache.txt)
[gcc-LastTest.log.gz](https://github.com/ginkgo-project/ginkgo/files/10911542/gcc-LastTest.log.gz)
[nvidia-CMakeCache.txt](https://github.com/ginkgo-project/ginkgo/files/10911543/nvidia-CMakeCache.txt)
[nvidia-LastTest.log.gz](https://github.com/ginkgo-project/ginkgo/files/10911544/nvidia-LastTest.log.gz)

**Comment by yhmtsai** [comment]
Thanks for reporting it! We also found the cusparse csr issue in the #1212 .
about more failed test from the nvidia compiler, it might be from too tight tolerance.
Is it consistent when running several times?

**Comment by upsj** [comment]
The cuSPARSE bug in `matrix_cuda` only concerns strided (non-packed) vectors, so even if you were using CSR storage, this would most likely not cause any issues for you.

**Comment by yhmtsai** [comment]
@pwxy Could you try this branch https://github.com/ginkgo-project/ginkgo/tree/spmm_single_vector to check whether pass matrix_cuda or not?
It only fallbacks to our own implementation when sparselib with single strided vector

**Comment by yhmtsai** [comment]
gcc version: it should be solved with #1322 
nvhpc: Dense and LU error should be solved with #1306, and CsrSparseLib should also be solved with #1322
However, the issues from Cgs, math, Jacobi are not related to #1306 and #1322
the math test `gko::is_finite(limits::quiet_NaN())` returns `true` not `false`.
I guess nvhpc by default uses fastmath such that nan is not actual nan.
maybe disable fast math can also solve the left issues.

**Comment by thoasm** [comment]
We use this test for both host and device:
```
return abs(value) < infinity;
```
--- https://github.com/ginkgo-project/ginkgo/blob/develop/include/ginkgo/core/base/math.hpp#L1205
So we don't use `std::isfinite` on host (or device).

**Comment by tcojean** [comment]
Should be fixed after #1331. Please reopen if that is not the case.
